sdk/python/kfp/dsl/_base_op.py

```diff
@@ -0,0 +1,781 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+import warnings
+from typing import Any, List, TypeVar, Union, Callable
+from kubernetes.client.models import (
+    V1Container, V1EnvVar, V1EnvFromSource, V1SecurityContext, V1Probe,
+    V1ResourceRequirements, V1VolumeDevice, V1VolumeMount, V1ContainerPort,
+    V1Lifecycle)
+
+from . import _pipeline
+from . import _pipeline_param
+
+# generics
+T = TypeVar('T')
+# type alias: either a string or a list of string
+StringOrStringList = Union[str, List[str]]
+
+
+# util functions
+def deprecation_warning(func: Callable, op_name: str,
+                        container_name: str) -> Callable:
+    """Decorator function to give a pending deprecation warning"""
+
+    def _wrapped(*args, **kwargs):
+        warnings.warn(
+            '`dsl.ContainerOp.%s` will be removed in future releases. '
+            'Use `dsl.ContainerOp.container.%s` instead.' %
+            (op_name, container_name), PendingDeprecationWarning)
+        return func(*args, **kwargs)
+
+    return _wrapped
+
+
+def as_list(value: Any, if_none: Union[None, List] = None) -> List:
+    """Convert any value except None to a list if not already a list."""
+    if value is None:
+        return if_none
+    return value if isinstance(value, list) else [value]
+
+
+def create_and_append(current_list: Union[List[T], None], item: T) -> List[T]:
+    """Create a list (if needed) and appends an item to it."""
+    current_list = current_list or []
+    current_list.append(item)
+    return current_list
+
+
+class Container(V1Container):
```

Comment:
Could you move Container to its own module in _container.py?

Reply:
Sure, but this needs some extra discussion on what to do with the `Sidecar` class.

Since moving `Container` is an orthogonal change to this PR, can we not block merging this PR for this, and get to moving `Container` to its new place right after this gets merged?

Reply:
I am fine to have another PR for it.

Reply:
OK, we will follow up with a PR and tidy up the code right after this one has been merged.

Reply:
AFAIK, ResourceOps cannot use containers or Sidecars, so it might have been better to leave them in the `_container_op.py`. This will make the diff size significantly smaller. Right now I do not know what changes were made to these classes.
Moving things around also breaks git blame.

Reply:
I'm fixing this issue today.

Reply:
Done. Can you please merge the following PR into your branch?: https://github.com/arrikto/kubeflow-pipelines/pull/1 It's 1 button click on the linked GitHub page.

Reply:
Done.
I've opened a PR into your branch: https://github.com/arrikto/kubeflow-pipelines/pull/2
Waiting for the e2e tests to pass (the unit tests are passing).

After the tests pass (https://prow.k8s.io/pr-history/?org=kubeflow&repo=pipelines&pr=1222) It's just 1 click on GitHub to merge the PR into your branch and then we can merge this PR to master.

Does that sound OK to you?

We really want to merge this PR today or tomorrow morning.

Reply:
Hi @Ark-kun!

What you are saying totally makes sense. I just approved your PR.
We will continue working on the rest of the things we have to do for this PR.

Thank you for taking the time to fix this!